File: agbac-min/agents/java/InSessionTokenRequest.java
package com.agbac.token;

import com.agbac.sender.HybridSender;
import com.agbac.adapters.BaseAdapter;

import java.util.Map;

/**
 * Handles OAuth token requests for in-session agents.
 * Ensures that the token includes both:
 *  - sub: the agent identity
 *  - act: the human identity (from session via HybridSender)
 */
public class InSessionTokenRequest {

    private final HybridSender sender;
    private final BaseAdapter adapter;

    public InSessionTokenRequest(HybridSender sender, BaseAdapter adapter) {
        this.sender = sender;
        this.adapter = adapter;
    }

    /**
     * Request an OAuth token for an in-session agent.
     *
     * @param sessionData Authenticated user session containing human info
     * @param agentClientId Agent client ID (sub)
     * @return access token string
     */
    public String requestToken(Map<String, Object> sessionData, String agentClientId) {
        try {
            Map<String, Object> actData = sender.provideActForInSession(sessionData);

            // Prepare token request payload
            Map<String, Object> tokenPayload = Map.of(
                    "client_id", agentClientId,
                    "grant_type", "client_credentials",
                    "act", actData
            );

            // Call adapter to request token from IAM
            String token = adapter.requestToken(tokenPayload);

            log("In-session token obtained for agent " + agentClientId + " with act " + actData);
            return token;

        } catch (Exception e) {
            log("Failed to request in-session token: " + e.getMessage());
            throw new RuntimeException("In-session token request failed", e);
        }
    }

    private void log(String message) {
        System.out.println("[InSessionTokenRequest] " + message);
    }
}

File: agbac-min/agents/typescript/insession/inSessionTokenRequest.ts
/**
 * inSessionTokenRequest.ts
 * Handles OAuth token request for in-session agents, including the human `act` claim.
 * Works with HybridSender to include `act` data in token request.
 */

import axios from 'axios';
import winston from 'winston';
import { HybridSender, HumanActor } from './hybridSender';
import { BaseAdapter } from './adapters/baseAdapter';

// Configure logging
const logger = winston.createLogger({
    level: 'info',
    format: winston.format.combine(
        winston.format.timestamp(),
        winston.format.printf(({ timestamp, level, message }) => `${timestamp} [${level}]: ${message}`)
    ),
    transports: [new winston.transports.Console()]
});

export class InSessionTokenRequest {
    private hybridSender: HybridSender;
    private adapter: BaseAdapter;

    constructor(hybridSender: HybridSender, adapter: BaseAdapter) {
        this.hybridSender = hybridSender;
        this.adapter = adapter;
    }

    /**
     * Request an OAuth token for an in-session agent including the human `act` claim
     */
    async requestToken(): Promise<string> {
        try {
            const actData: HumanActor = this.hybridSender.getInSessionAct();
            logger.info(`Preparing token request including human ACT: ${actData.sub}`);

            // Build token request payload according to adapter
            const payload = this.adapter.buildTokenRequestPayload(actData);

            // Execute token request
            const tokenResponse = await axios.post(this.adapter.tokenUrl, payload, {
                headers: this.adapter.getTokenRequestHeaders()
            });

            const accessToken = tokenResponse.data.access_token;
            logger.info(`Token request successful for in-session agent`);
            return accessToken;

        } catch (error) {
            logger.error(`In-session token request failed: ${error.message || error}`);
            throw error;
        }
    }
}
